File: springboot-02-ioccontainer/src/main/java/com/linkedbear/springboot/assemble/d_spi/SpringSpiApplication.java
package com.linkedbear.springboot.assemble.d_spi;

import com.linkedbear.springboot.assemble.d_spi.bean.DemoDao;
import org.springframework.core.io.support.SpringFactoriesLoader;

import java.util.List;

public class SpringSpiApplication {
    
    public static void main(String[] args) throws Exception {
        List<DemoDao> demoDaos = SpringFactoriesLoader
                .loadFactories(DemoDao.class, SpringSpiApplication.class.getClassLoader());
        demoDaos.forEach(System.out::println);
    
        System.out.println("------------------------------------------------");
    
        List<String> daoClassNames = SpringFactoriesLoader
                .loadFactoryNames(DemoDao.class, SpringSpiApplication.class.getClassLoader());
        daoClassNames.forEach(System.out::println);
    }
}

File: springboot-02-ioccontainer/src/main/kotlin/com/linkedbear/springboot/assemble/d_spi/SpringSpiApplication.kt
package com.linkedbear.springboot.assemble.d_spi

import com.linkedbear.springboot.assemble.d_spi.bean.DemoDao
import org.springframework.core.io.support.SpringFactoriesLoader

class SpringSpiApplication {
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val demoDaos = SpringFactoriesLoader.loadFactories(
                DemoDao::class.java,
                SpringSpiApplication::class.java.classLoader,
            )
            demoDaos.forEach {
                println(it)
            }
            println("-------------------------------")
            val daoClassNames = SpringFactoriesLoader.loadFactoryNames(
                DemoDao::class.java,
                SpringSpiApplication::class.java.classLoader,
            )
            daoClassNames.forEach {
                println(it)
            }
        }
    }
}